core/bin/snapshots_creator/src/main.rs

```diff
@@ -0,0 +1,304 @@
+        .access_storage_tagged("snapshots_creator")
+        .await?;
+
+    // we subtract 1 so that after restore, EN node has at least one l1 batch to fetch
+    let l1_batch_number = conn.blocks_dal().get_sealed_l1_batch_number().await? - 1;
+
+    let mut master_conn = master_pool
+        .access_storage_tagged("snapshots_creator")
+        .await?;
+    if master_conn
+        .snapshots_dal()
+        .get_snapshot_metadata(l1_batch_number)
+        .await?
+        .is_some()
+    {
+        tracing::info!("Snapshot for L1 batch number {l1_batch_number} already exists, exiting",);
+        return Ok(());
+    }
+    drop(master_conn);
+
+    let last_miniblock_number_in_batch = conn
+        .blocks_dal()
+        .get_miniblock_range_of_l1_batch(l1_batch_number)
+        .await?
+        .context("Error fetching last miniblock number")?
+        .1;
+    let distinct_storage_logs_keys_count = conn
+        .snapshots_creator_dal()
+        .get_distinct_storage_logs_keys_count(l1_batch_number)
+        .await?;
+
+    drop(conn);
+
+    let chunk_size = config.storage_logs_chunk_size;
+    // we force at least 10 chunks to avoid situations where only one chunk is created in tests
+    let chunks_count = max(10, ceil_div(distinct_storage_logs_keys_count, chunk_size));
+
+    METRICS.storage_logs_chunks_count.set(chunks_count);
+
+    tracing::info!(
+        "Creating snapshot for storage logs up to miniblock {last_miniblock_number_in_batch}, l1_batch {}",
+        l1_batch_number.0
+    );
+    tracing::info!("Starting to generate {chunks_count} chunks of expected size {chunk_size}");
+
+    let factory_deps_output_file = process_factory_deps(
+        &*blob_store,
+        &replica_pool,
+        last_miniblock_number_in_batch,
+        l1_batch_number,
+    )
+    .await?;
+
+    let mut storage_logs_output_files = vec![];
+
+    METRICS
+        .storage_logs_chunks_left_to_process
+        .set(chunks_count);
+    let mut tasks =
+        FuturesUnordered::<Pin<Box<dyn Future<Output = anyhow::Result<String>>>>>::new();
+    let mut last_chunk_id = 0;
+    while last_chunk_id < chunks_count || !tasks.is_empty() {
+        while (tasks.len() as u32) < config.concurrent_queries_count && last_chunk_id < chunks_count
+        {
+            tasks.push(Box::pin(process_storage_logs_single_chunk(
+                &*blob_store,
+                &replica_pool,
+                last_miniblock_number_in_batch,
+                l1_batch_number,
+                last_chunk_id,
+                chunks_count,
+            )));
+            last_chunk_id += 1;
+        }
+        if let Some(result) = tasks.next().await {
+            tracing::info!(
+                "Completed chunk {}/{}, {} chunks are still in progress",
+                last_chunk_id - tasks.len() as u64,
+                chunks_count,
+                tasks.len()
+            );
+            storage_logs_output_files.push(result.context("Chunk task failed")?);
```

Comment:
Just to clarify: Is it intended that `storage_logs_output_files` can be ordered in any way (i.e., not by `chunk_id`)?

Reply:
definitely not! nice catch